File: src/livekit_agent_simulator_rust/crates/lks-core/src/script/summary.rs
//! Post-run caller behavior aggregates — byte-parity port of `script/summary.py`.
//! Always safe to call; zeros when the run had no scripted caller behavior.

use serde_json::{json, Map, Value as Json};

use super::counts_for_recovery_barge;

fn ev_kind(e: &Map<String, Json>) -> &str {
    e.get("kind").and_then(|v| v.as_str()).unwrap_or("")
}

fn ev_spec(e: &Map<String, Json>) -> &Map<String, Json> {
    static EMPTY: std::sync::OnceLock<Map<String, Json>> = std::sync::OnceLock::new();
    e.get("spec")
        .and_then(|v| v.as_object())
        .unwrap_or_else(|| EMPTY.get_or_init(Map::new))
}

fn ev_mono(e: &Map<String, Json>) -> i64 {
    e.get("ts_mono_ms").and_then(|v| v.as_i64()).unwrap_or(0)
}

/// Aggregate barge / silence / recovery stats for summary.json + report player.
pub fn build_caller_behavior_summary(events: &[Map<String, Json>]) -> Map<String, Json> {
    let cues: Vec<&Map<String, Json>> = events
        .iter()
        .filter(|e| ev_kind(e) == "sim.script.cue")
        .collect();
    let waits: Vec<&Map<String, Json>> = events
        .iter()
        .filter(|e| ev_kind(e) == "sim.script.wait")
        .collect();

    let mut barges: Vec<&Map<String, Json>> = Vec::new();
    let mut by_class: Map<String, Json> = Map::new();
    for e in &cues {
        let spec = ev_spec(e);
        let cls = spec
            .get("class")
            .or_else(|| spec.get("interrupt_class"))
            .map(|v| v.as_str().unwrap_or(&v.to_string()).to_string())
            .filter(|s| !s.is_empty());
        if let Some(c) = &cls {
            let n = by_class.get(c).and_then(|v| v.as_i64()).unwrap_or(0) + 1;
            by_class.insert(c.clone(), json!(n));
        }
        if counts_for_recovery_barge(
            spec.get("barge_in")
                .and_then(|v| v.as_bool())
                .unwrap_or(false),
            cls.as_deref(),
        ) {
            barges.push(e);
        }
    }
    let barges_during: Vec<&&Map<String, Json>> = barges
        .iter()
        .filter(|e| {
            ev_spec(e)
                .get("during_agent_speech")
                .and_then(|v| v.as_bool())
                .unwrap_or(false)
        })
        .collect();
    let cues_during: Vec<&&Map<String, Json>> = cues
        .iter()
        .filter(|e| {
            ev_spec(e)
                .get("during_agent_speech")
                .and_then(|v| v.as_bool())
                .unwrap_or(false)
        })
        .collect();
    let silence_events: Vec<&Map<String, Json>> = events
        .iter()
        .filter(|e| ev_kind(e) == "silence.detected")
        .collect();
    let interruptions: Vec<&Map<String, Json>> = events
        .iter()
        .filter(|e| ev_kind(e) == "interruption")
        .collect();
    let agent_finals: Vec<&Map<String, Json>> = events
        .iter()
        .filter(|e| ev_kind(e) == "transcript.agent.final")
        .collect();

    let barge_ms = barges.first().map(|e| ev_mono(e));
    let silence_ms = waits.first().map(|e| ev_mono(e));

    let mut agent_after_barge: i64 = 0;
    let mut recovery_ms: Option<i64> = None;
    if let Some(bms) = barge_ms {
        let after: Vec<i64> = agent_finals
            .iter()
            .filter(|e| ev_mono(e) > bms)
            .map(|e| ev_mono(e))
            .collect();
        agent_after_barge = after.len() as i64;
        if let Some(first) = after.first() {
            recovery_ms = Some(first - bms);
        }
    }

    let agent_after_silence = match silence_ms {
        Some(sms) => agent_finals.iter().filter(|e| ev_mono(e) >= sms).count() as i64,
        None => 0,
    };

    let mut assets: Vec<String> = Vec::new();
    for e in events {
        let k = ev_kind(e);
        if k != "sim.script.cue" && k != "sim.script_inject" {
            continue;
        }
        if let Some(a) = ev_spec(e).get("asset") {
            let s = a.as_str().unwrap_or(&a.to_string()).to_string();
            if !s.is_empty() && !assets.contains(&s) {
                assets.push(s);
            }
        }
    }

    let mut out = Map::new();
    out.insert("script_cues_fired".into(), json!(cues.len()));
    out.insert("waits_fired".into(), json!(waits.len()));
    out.insert("barges_fired".into(), json!(barges.len()));
    out.insert("barges_during_agent".into(), json!(barges_during.len()));
    out.insert("cues_during_agent".into(), json!(cues_during.len()));
    out.insert("silences_held".into(), json!(waits.len()));
    out.insert("silence_events".into(), json!(silence_events.len()));
    out.insert("interruptions".into(), json!(interruptions.len()));
    out.insert("agent_finals_after_barge".into(), json!(agent_after_barge));
    out.insert(
        "agent_finals_after_silence".into(),
        json!(agent_after_silence),
    );
    out.insert(
        "recovery_ms".into(),
        recovery_ms.map(|v| json!(v)).unwrap_or(Json::Null),
    );
    out.insert(
        "cue_assets".into(),
        Json::Array(assets.into_iter().map(Json::String).collect()),
    );
    out.insert("by_class".into(), Json::Object(by_class));
    out
}

File: src/livekit_agent_simulator/script/summary.py
"""Post-run caller behavior aggregates for summary.json / web."""

from __future__ import annotations

from typing import Any

def build_caller_behavior_summary(events: list[dict]) -> dict[str, Any]:
    """Aggregate barge / silence / recovery stats for summary.json + report player.

    Always safe to call; zeros when the run had no scripted caller behavior.
    """
    cues = [e for e in events if e.get("kind") == "sim.script.cue"]
    waits = [e for e in events if e.get("kind") == "sim.script.wait"]
    # Every script action kind that counts as "a cue fired" (speak, DTMF tone,
    # wait hold, hang-up). DTMF steps emit sim.script.dtmf, so a pure-DTMF run
    # must not report 0 cues fired.
    script_action_kinds = {
        "sim.script.cue",
        "sim.script.dtmf",
        "sim.script.wait",
        "sim.script.hang_up",
    }
    all_script_actions = [e for e in events if e.get("kind") in script_action_kinds]
    from .models import counts_for_recovery_barge

    barges = []
    by_class: dict[str, int] = {}
    for e in cues:
        spec = e.get("spec") or {}
        cls = str(spec.get("class") or spec.get("interrupt_class") or "") or None
        if cls:
            by_class[cls] = by_class.get(cls, 0) + 1
        if counts_for_recovery_barge(
            barge_in=bool(spec.get("barge_in")),
            interrupt_class=cls,
        ):
            barges.append(e)
    barges_during = [
        e for e in barges if (e.get("spec") or {}).get("during_agent_speech")
    ]
    cues_during = [
        e for e in cues if (e.get("spec") or {}).get("during_agent_speech")
    ]
    silence_events = [e for e in events if e.get("kind") == "silence.detected"]
    interruptions = [e for e in events if e.get("kind") == "interruption"]
    agent_finals = [e for e in events if e.get("kind") == "transcript.agent.final"]

    barge_ms = barges[0].get("ts_mono_ms") if barges else None
    silence_ms = waits[0].get("ts_mono_ms") if waits else None

    agent_after_barge = 0
    recovery_ms: int | None = None
    if barge_ms is not None:
        after = [
            int(e.get("ts_mono_ms") or 0)
            for e in agent_finals
            if int(e.get("ts_mono_ms") or 0) > int(barge_ms)
        ]
        agent_after_barge = len(after)
        if after:
            recovery_ms = after[0] - int(barge_ms)

    agent_after_silence = 0
    if silence_ms is not None:
        agent_after_silence = sum(
            1
            for e in agent_finals
            if int(e.get("ts_mono_ms") or 0) >= int(silence_ms)
        )

    assets: list[str] = []
    for e in events:
        if e.get("kind") not in ("sim.script.cue", "sim.script_inject"):
            continue
        a = (e.get("spec") or {}).get("asset")
        if a and str(a) not in assets:
            assets.append(str(a))

    return {
        "script_cues_fired": len(all_script_actions),
        "waits_fired": len(waits),
        "barges_fired": len(barges),
        "barges_during_agent": len(barges_during),
        "cues_during_agent": len(cues_during),
        "silences_held": len(waits),
        "silence_events": len(silence_events),
        "interruptions": len(interruptions),
        "agent_finals_after_barge": agent_after_barge,
        "agent_finals_after_silence": agent_after_silence,
        "recovery_ms": recovery_ms,
        "cue_assets": assets,
        "by_class": by_class,
    }


